packages/admin/cms-admin/src/blocks/table/ActionsCell.tsx

```diff
@@ -49,6 +52,28 @@ export const ActionsCell = ({ row, updateState }: Props) => {
         });
     };
 
+    const duplicateRow = () => {
+        updateState((state) => {
+            const currentRowIndex = state.rows.findIndex(({ id }) => id === row.id);
+            const rowToDuplicate = state.rows[currentRowIndex];
+
+            if (!rowToDuplicate) {
+                snackbarApi.showSnackbar(
+                    <Snackbar autoHideDuration={5000}>
+                        <Alert severity="error">
+                            <FormattedMessage id="comet.tableBlock.failedToDuplicateRow" defaultMessage="Failed to duplicate row" />
+                        </Alert>
+                    </Snackbar>,
+                );
+
+                return state;
+            }
```

Comment:
IMO this should be outside the updateState call. updateState is basically a React setState call, and I wouldn't add a side effect to it.

Reply:
Do you mean only the `<Snackbar />` part or the whole `if`? 

The variable checked `rowToDuplicate` is dependent on the `state`, so we'd need to pass in the `state` but I'd prefer not to use the outside `state`, as that has a (probably very small) chance of not being up-to-date when `updateState` is called 🤔

Reply:
> Do you mean only the <Snackbar /> part or the whole if?

The whole if

> The variable checked rowToDuplicate is dependent on the state, so we'd need to pass in the state but I'd prefer not to use the outside state, as that has a (probably very small) chance of not being up-to-date when updateState is called

Yes, that's what would be better. It's odd that ActionCell has an updateState prop but no state prop anyway.

Possible alternative: Add an `onDuplicateRow` prop to ActionCell and handle the state updates in the parent component. This would probably be the best way.

Reply:
I'll pass in `state` for now, will maybe refactor this for testability anyway 🙂

Reply:
Done in f90dc815933e7180b33e3f441db5106f678100db